hudi-common/src/main/java/org/apache/hudi/common/util/hash/BucketIndexUtil.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.common.util.hash;
+
+import org.apache.hudi.common.util.Functions;
+
+/**
+ * Utility class for bucket index.
+ */
+public class BucketIndexUtil {
+
+  /**
+   * This method is used to get the partition index calculation function of a bucket.
+   * "partition.hashCode() / (parallelism / bucketNum) * bucketNum" divides the parallelism into
+   * sub-intervals of length bucket_num, different partitions will be mapped to different sub-interval,
+   * ensure that the data across multiple partitions is evenly distributed.
+   *
+   * @param bucketNum   Bucket number per partition
+   * @param parallelism Parallelism of the task
+   * @return The partition index of this bucket.
+   */
+  public static Functions.Function2<String, Integer, Integer> getPartitionIndexFunc(int bucketNum, int parallelism) {
+    return (partition, curBucket) -> {
+      int partitionIndex = (partition.hashCode() & Integer.MAX_VALUE) % parallelism * bucketNum;
+      int globalIndex = partitionIndex + curBucket;
```

Comment:
The logic in Flink has got a regression?

Reply:
It's not sure, no problem found in my local area, I re -triggered
<img width="1412" alt="image" src="https://github.com/apache/hudi/assets/20125927/c275edcc-62f0-49f0-9e53-a0e5c5eb0de7">

Reply:
after trying it many times, can't come up with this problem

Reply:
Did you check the original commit PR description and comments, maybe we can sync up there first: https://github.com/apache/hudi/pull/11346

Reply:
yes, I have check, added more complete ut.

Reply:
- avg: total bucket file / parallelism, in this case, total bucket number is 9 * bucketNumber
- parallelism: task can max running parallelism
- bucketNumber:
- inlimit: match algorithm condition, fluctuate in the avg
- outOflimit: 
- real run parallelism size: the `partitionIndexFunc` value, which is meat work index, and why it will bigger than parallelism in flink old logical, because it can be negative.

use flink old logical, only the parallelism is twice bigger than bucketNumber work will.

avg : 18
parallelism : 6
bucketNumber : 12
inlimit : [18]
outOfLimit : [10, 8, 10, 8, 10, 8, 10, 8, 10, 8]
real run parallelism size : 11
==================================
avg : 21
parallelism : 5
bucketNumber : 12
inlimit : [18]
outOfLimit : [8, 12, 14, 12, 14, 8, 11, 11]
real run parallelism size : 9
==================================
avg : 4
parallelism : 24
bucketNumber : 12
inlimit : [4, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2]
outOfLimit : []
real run parallelism size : 47
==================================
avg : 4
parallelism : 23
bucketNumber : 12
inlimit : [5, 4, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 2, 3, 3, 3, 2, 3, 2, 3, 2]
outOfLimit : []
real run parallelism size : 45
==================================
avg : 0
parallelism : 100
bucketNumber : 5
inlimit : []
outOfLimit : [4, 4, 4, 4, 4, 5, 5, 5, 5, 5]
real run parallelism size : 10
==================================

after use fix logical in this pr:

avg : 18
parallelism : 6
bucketNumber : 12
inlimit : [18, 18, 18, 18, 18, 18]
outOfLimit : []
real run parallelism size : 6
==================================
avg : 21
parallelism : 5
bucketNumber : 12
inlimit : [21, 21, 22, 22, 22]
outOfLimit : []
real run parallelism size : 5
==================================
avg : 4
parallelism : 24
bucketNumber : 12
inlimit : [5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4]
outOfLimit : []
real run parallelism size : 24
==================================
avg : 4
parallelism : 23
bucketNumber : 12
inlimit : [4, 5, 5, 5, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4]
outOfLimit : []
real run parallelism size : 23
==================================
avg : 0
parallelism : 100
bucketNumber : 5
inlimit : [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
outOfLimit : []
real run parallelism size : 45
==================================

Reply:
@xicm Can you help to confirm here?

Reply:
Did you test with multi paritions in one write?  #11346 is a fix for data skew when a write contains multiple partitions.  details https://github.com/apache/hudi/issues/11288

Reply:
@xicm the ut is multi partitions 
```scala
private void initData(Map<Integer, Integer> parallelism2TaskCount, int bucketNumber,
                               Functions.Function2<String, Integer, Integer> partitionIndexFunc) {
    parallelism2TaskCount.clear();
    Arrays.asList("2021-01-01", "2021-01-02", "2021-01-03", "2021-01-04",
        "2022-01-05", "2022-01-06", "2022-01-07", "2022-01-08", "2022-01-09").forEach(partition -> {
          for (int bucketIndex = 0; bucketIndex < bucketNumber; bucketIndex++) {
            putIndexCount(parallelism2TaskCount, partitionIndexFunc.apply(partition, bucketIndex));
          }
        });
  }
}
```

Reply:
@xicm this reminds me that I will add a test in the non-partition

Reply:
Dose inlimit means how many piece of data a task deal with?
This case,
```
avg : 4
parallelism : 23
bucketNumber : 12
inlimit : [4, 5, 5, 5, 6, 6, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 4, 4, 4, 4, 4, 4, 4]
outOfLimit : []
real run parallelism size : 23
```

there are two tasks dealing with 6 pieces of data. 
I think we expect a few task dealing with 4, a few dealing with 5.

Reply:
@xicm since tasks are not uniformly assigned by a single node, it is unlikely to control data allocation with such fine granularity. Instead, the goal is to distribute the data as evenly as possible.

Reply:
I find a corner case,
parallelism=13, bucketNum=5, and test with partition "2021-01-01", "2021-01-03".
inlimit : [2, 1, 1, 1, 1, 1, 1, 2]

Reply:
@xicm look like it's hard to cover all case, because `partition value` hash code is hard to predict. If there isn't a better algorithm, we can proceed based on practical considerations. In most scenarios, partitions appear consecutively. The current fixed algorithm is relatively stable.

Reply:
@danny0405 what do you think about?

Reply:
Can you summarize the current drawback of the Flink partitioner algorithm for 3 scenarios and what is the main reason to rollback the changes to the original?

Reply:
These two algorithms divide the parallelism into bucket-sized blocks, so that the tasks are evenly distributed.

(partition.hashCode() & Integer.MAX_VALUE) % parallelism * bucketNum means the start index of the buckets in a partition.

The difference is whether “% parallelism” or “/ (parallelism/bucketNum)”。  

Assume bucketNum = 2, parallelism = 10

“(partition.hashCode() & Integer.MAX_VALUE) % parallelism” means the start index will be any value in [0-parallelism ],  diffrent partitions the value can be next to each other。

![image](https://github.com/apache/hudi/assets/36392121/446c4c32-8854-42fb-aaf2-41bf5bd0d9b1)


“(partition.hashCode() & Integer.MAX_VALUE) / (parallelism / bucketNum )” ，There will be a interval between the start index of different partitions

![image](https://github.com/apache/hudi/assets/36392121/27d17c57-5eab-4afb-bf9c-20125cf4a8bd)

Reply:
> Can you summarize the current drawback of the Flink partitioner algorithm for 3 scenarios and what is the main reason to rollback the changes to the original?

no, the original logical in flink is `(partition.hashCode() & Integer.MAX_VALUE) % parallelism` current is `(partition.hashCode() & Integer.MAX_VALUE) % parallelism * bucketNumber`

Reply:
@xicm `(partition.hashCode() & Integer.MAX_VALUE) % parallelism` is right, but `(partition.hashCode() & Integer.MAX_VALUE) / (parallelism / bucketNum + 1) * bucketNum` have a little problem, because it is `/` not `%`, so I think it should be like this. Each adjacent partition will be in the same slot, the range depend on `parallelism / bucketNum`, like you case, is `5`
![image](https://github.com/apache/hudi/assets/20125927/c93670c0-9ef2-4f91-b401-3b60c9955fc6)

Reply:
>  Each adjacent partition will be in the same slot
this is not what we expect.

you can run the test case with parallelism = 10, bucketNumber = 5 and partition = ["2021-01-01", "2021-01-03"], the result is inlimit : [2, 2, 2, 2, 2], half of TM are idle.

Reply:
@xicm hi, https://github.com/apache/hudi/pull/11578#discussion_r1668707261, as the comment say, it can not cover all case, just based on practical considerations